Fix broken git gateway github integration

melbourne2991 opened the pull request:
Closes #2629

**Summary**
Fixes regression described in the issue above.

**Test plan**
Manually tested.

barthc commented on packages/netlify-cms-backend-github/src/implementation.js:
```diff
@@ -85,7 +86,7 @@ export default class GitHub {
     var repoExists = false;
     while (!repoExists) {
-      repoExists = await fetch(`${this.api_root}/repos/${repo}`, {
+      repoExists = await fetch(`${this.api_root}${this.repoURLPrefix}${repo}`, {
```
I don't think the `git-gateway` backend will be calling this.

barthc commented on packages/netlify-cms-backend-github/src/implementation.js:
```diff
@@ -114,7 +115,7 @@ export default class GitHub {
     if (!this._userIsOriginMaintainerPromises[username]) {
       this._userIsOriginMaintainerPromises[username] = fetch(
-        `${this.api_root}/repos/${this.originRepo}/collaborators/${username}/permission`,
+        `${this.api_root}${this.repoURLPrefix}${this.originRepo}/collaborators/${username}/permission`,
```
Also, this won't be called.

barthc commented on packages/netlify-cms-backend-github/src/implementation.js:
```diff
@@ -142,7 +143,7 @@ export default class GitHub {
     await getPermissionToFork();
 
-    const fork = await fetch(`${this.api_root}/repos/${this.originRepo}/forks`, {
+    const fork = await fetch(`${this.api_root}${this.repoURLPrefix}${this.originRepo}/forks`, {
```
And here as well.

melbourne2991 replied:
Yep so `this.options.proxied` will be false and this will still evaluate `/repos/` - I don't think this should be a problem? Is there any benefit to having `/repos/` hardcoded in the method?

barthc replied:
It's just for readability and debugging sake. With the `this.options.proxied` check, if I  am looking at the code, I am thinking the `git-gateway` is calling those APIs.

barthc commented:
I think a better way to fix this issue is to DRY the `this.useOpenAuthoring ? this.originRepoURL : this.repoURL` check inside the `API.js` `constructor`  eliminating the need for the same check inside `implementation.js`, @Benaiah  thoughts?

erquhart commented:
@melbourne2991 are you trying to use Open Authoring with Git Gateway?

erezrokah commented:
> I think a better way to fix this issue is to DRY the `this.useOpenAuthoring ? this.originRepoURL : this.repoURL` check inside the `API.js` `constructor` eliminating the need for the same check inside `implementation.js`, @Benaiah thoughts?

I think `this.useOpenAuthoring ? this.originRepoURL : this.repoURL` is not required anymore in most places (if not all of them), since in the GraphQL pr I made sure `this.originRepoURL === this.repoURL` when `useOpenAuthoring === false`. Meaning one can just use `this.originRepoURL`.

melbourne2991 commented:
@erquhart Nope, Git gateway is broken if you are using Github. The regression is because of the open authoring changes. 

What has happened is you have the Git gateway backend which [creates its own instance of the Github backend](https://github.com/netlify/netlify-cms/blob/master/packages/netlify-cms-backend-git-gateway/src/implementation.js#L161) and passes in an override for the "API" instance which extends from the Github API instance and, importantly, [overrides the repoURL instance variable](https://github.com/netlify/netlify-cms/blob/master/packages/netlify-cms-backend-git-gateway/src/GitHubAPI.js#L10).

The regression is that the GithubAPI was updated to allow overriding of the repoURL on a per method basis, and the [Github backend was updated to pass through a repoURL](https://github.com/netlify/netlify-cms/pull/2430/files#diff-551fa63580f74f0768649609163f3005R201]) that had `/repos/` hard coded in it, so what that means is even though we [override the repoURL here for git gateway](https://github.com/netlify/netlify-cms/blob/master/packages/netlify-cms-backend-git-gateway/src/GitHubAPI.js#L10) that override is now itself overriden by Github backend which is used by git gateway.

The fix is to update the per method override in Git backend so that it passes through a correct url, which means passing a url without `/repos/` for Git gateway.

Hopefully that all makes sense :sweat_smile:

erquhart commented:
Ah, didn't realize the hardcoded `/repos/` was just introduced. Can you update to use the approach mentioned by @erezrokah?

melbourne2991 commented:
I removed instances the repoURL that was being passed into API (I suppose we can remove it from the method signatures as well?). I did get a new error: `text is not a function` which is why I updated `fetchBlobContent`.

@erezrokah If you could take a quick look and let me know if those changes seem ok? Cheers

erezrokah commented:
@melbourne2991 I'll try to reproduce your issue locally first so I understand the problem better. Then I can evaluate the fix

erezrokah commented:
@melbourne2991 The current fix breaks the open authoring feature since when enabling open authoring files should always be read from the origin repo to reflect recent changes.

I've made an updated fix [here](https://github.com/erezrokah/netlify-cms/commit/dcb90185e15dfebd2e542fba962759b9d13776a6) which I believe solves both issues (also includes a fix to the GraphQL API usage with open authoring).

I've tested it both with `git-gateway` and REST/GraphQL open authoring, but it would be great if you can port it to your PR and verify as well.

melbourne2991 commented:
@erezrokah Works fine for me, nice one! (updated my commit)

erezrokah commented on packages/netlify-cms-backend-github/src/implementation.js:
```diff
@@ -221,30 +221,30 @@ export default class GitHub {
     const files = await this.api.listFiles(collection.get('folder'), { repoURL });
     const filteredFiles = files.filter(file => file.name.endsWith('.' + extension));
-    return this.fetchFiles(filteredFiles, { repoURL });
+    return this.fetchFiles(filteredFiles);
```
With the latest fix this should be `return this.fetchFiles(filteredFiles, { repoURL });`

erezrokah commented on packages/netlify-cms-backend-github/src/implementation.js:
```diff
@@ -221,30 +221,30 @@ export default class GitHub {
       label: collectionFile.get('label'),
     }));
-    return this.fetchFiles(files, { repoURL });
+    return this.fetchFiles(files);
```
Same as https://github.com/netlify/netlify-cms/pull/2631/files#r324445567

erezrokah commented on packages/netlify-cms-backend-github/src/implementation.js:
```diff
@@ -221,30 +221,30 @@ export default class GitHub {
           sem.take(() =>
             this.api
-              .readFile(file.path, file.sha, { repoURL })
+              .readFile(file.path, file.sha)
```
This should be `.readFile(file.path, file.sha, { repoURL })`

melbourne2991 replied:
Added those changes and retested on my end

erquhart commented:
Thanks @melbourne2991!

@erezrokah give a comment/approval if you're happy with the changes here.

erezrokah commented:
@erquhart Looks good to me
